runtime/ops/tls.rs

```diff
@@ -70,6 +85,421 @@ impl StoresClientSessions for ClientSessionMemoryCache {
+        && self.tls.wants_read()
+        && self.rd_shut != Shut::TcpShut
+      {
+        let mut wrapper = ImplementReadTrait(&mut self.tcp);
+        match self.tls.read_tls(&mut wrapper) {
+          Ok(0) if self.tls.is_handshaking() => {
+            let err =
+              io::Error::new(ErrorKind::UnexpectedEof, "tls handshake eof");
+            return Poll::Ready(Err(err));
+          }
+          Ok(0) => self.rd_shut = Shut::TcpShut,
+          Ok(_) => match self.tls.process_new_packets() {
+            Ok(_) => {}
+            Err(err) => {
+              let err = io::Error::new(ErrorKind::InvalidData, err);
+              return Poll::Ready(Err(err));
+            }
+          },
+          Err(err) if err.kind() == ErrorKind::WouldBlock => {
+            match self.tcp.poll_read_ready(cx) {
+              Poll::Ready(result) => result?,
+              Poll::Pending => tcp_read_ready = false,
+            }
+          }
+          Err(err) => return Poll::Ready(Err(err)),
+        }
+      }
+
+      if tcp_write_ready && self.tls.wants_write() {
+        continue;
+      }
+
+      if for_reading
+        && !self.tls.is_handshaking()
+        && (!self.tls.wants_read() || self.rd_shut == Shut::TcpShut)
+      {
+        return Poll::Ready(Ok(()));
+      }
+      if for_writing && !self.tls.is_handshaking() && !self.tls.wants_write() {
+        return Poll::Ready(Ok(()));
+      }
+
+      return Poll::Pending;
+    }
+  }
+}
+
+impl AsyncRead for TlsStream {
+  fn poll_read(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &mut ReadBuf<'_>,
+  ) -> Poll<io::Result<()>> {
+    ready!(self.poll_io(cx, true, false))?;
+
+    let buf_slice = unsafe {
+      &mut *(buf.unfilled_mut() as *mut [MaybeUninit<u8>] as *mut [u8])
+    };
+
+    match self.tls.read(buf_slice) {
+      Ok(bytes_read) => {
+        unsafe { buf.assume_init(bytes_read) };
+        buf.advance(bytes_read);
+        Poll::Ready(Ok(()))
+      }
+      Err(err) if err.kind() == ErrorKind::ConnectionAborted => {
+        // Rustls `read()` returns `ConnectionAborted` when it receives a
+        // `CloseNotify` alert; this indicates that the other end of the line
+        // has initiated a graceful shutdown at the TLS level.
+        self.rd_shut = Shut::TlsShut;
+        Poll::Ready(Ok(()))
+      }
+      Err(err) => Poll::Ready(Err(err)),
+    }
+  }
+}
+
+impl AsyncWrite for TlsStream {
+  fn poll_write(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &[u8],
+  ) -> Poll<io::Result<usize>> {
+    ready!(self.poll_io(cx, false, true))?;
+
+    match self.tls.write(buf) {
+      Ok(0) => unreachable!(),
+      Ok(bytes_written) => {
+        // We try to flush as much as we can, but since we have already handed
+        // off some bytes to rustls we can't return `Poll::Pending()` any more,
+        // as this would indicate to the caller that it should try again.
+        let _ = self.poll_io(cx, false, true)?;
```

Comment:
Yeah the number of bytes written gets discarded in that case; I don't think this is an issue because if it returns an error the socket is dead.

That said, with the right tactics I could be convinced to get rid of this `poll_io()` call entirely.